pkg/tempo/promsdprocessor/prom_sd_processor.go

```diff
@@ -0,0 +1,216 @@
+		logger:         util.Logger,
+	}, nil
+}
+
+func (p *promServiceDiscoProcessor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	rss := td.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		if rs.IsNil() {
+			continue
+		}
+		r := rs.Resource()
+		if r.IsNil() {
+			continue
+		}
+
+		p.processAttributes(r.Attributes())
+	}
+
+	return p.nextConsumer.ConsumeTraces(ctx, td)
+}
+
+func (p *promServiceDiscoProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: true}
+}
+
+// Start is invoked during service startup.
+func (p *promServiceDiscoProcessor) Start(_ context.Context, _ component.Host) error {
+	go p.watchServiceDiscovery()
+
+	go func() {
+		err := p.discoveryMgr.Run()
+		if err != nil {
+			level.Error(p.logger).Log("msg", "failed to start prom svc disco.  relabeling disabled", "err", err)
+		}
+	}()
+
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (p *promServiceDiscoProcessor) Shutdown(context.Context) error {
+	return nil
+}
+
+func (p *promServiceDiscoProcessor) watchServiceDiscovery() {
+	for targetGoups := range p.discoveryMgr.SyncCh() {
+		hostLabels := make(map[string]model.LabelSet)
+		level.Debug(p.logger).Log("msg", "syncing target groups", "count", len(targetGoups))
+		for jobName, groups := range targetGoups {
+			p.syncGroups(jobName, groups, hostLabels)
+		}
+		p.mtx.Lock()
+		p.hostLabels = hostLabels
+		p.mtx.Unlock()
+	}
+}
+
+func (p *promServiceDiscoProcessor) syncGroups(jobName string, groups []*targetgroup.Group, hostLabels map[string]model.LabelSet) {
+	level.Debug(p.logger).Log("msg", "syncing target group", "jobName", jobName)
+	for _, g := range groups {
+		p.syncTargets(jobName, g, hostLabels)
+	}
+}
+
+func (p *promServiceDiscoProcessor) syncTargets(jobName string, group *targetgroup.Group, hostLabels map[string]model.LabelSet) {
+	level.Debug(p.logger).Log("msg", "syncing targets", "count", len(group.Targets))
+
+	relabelConfig := p.relabelConfigs[jobName]
+	if relabelConfig == nil {
+		level.Warn(p.logger).Log("msg", "relabel config not found for job. skipping labeling", "jobName", jobName)
+		return
+	}
+
+	for _, t := range group.Targets {
+		discoveredLabels := group.Labels.Merge(t)
+
+		level.Debug(p.logger).Log("discoveredLabels", discoveredLabels)
+		var labelMap = make(map[string]string)
+		for k, v := range discoveredLabels.Clone() {
+			labelMap[string(k)] = string(v)
+		}
+		processedLabels := relabel.Process(labels.FromMap(labelMap), relabelConfig...)
+		level.Debug(p.logger).Log("processedLabels", processedLabels)
+
+		var labels = make(model.LabelSet)
+		for k, v := range processedLabels.Map() {
+			labels[model.LabelName(k)] = model.LabelValue(v)
+		}
+
+		address, ok := labels[addressLabel]
+		if !ok {
+			level.Warn(p.logger).Log("msg", "ignoring target, unable to find address", "labels", labels.String())
+			continue
+		}
```

Comment:
Just want to point out that this check is happening after relabeling, so a user could hackily change `__address__` to something else. Would that be a problem here?

Reply:
I don't think so.  Could be desirable if you needed to get something non-standard to work.